crates/sui-framework/packages/dwallet-system/sources/sui_state_proof.move

```diff
@@ -0,0 +1,145 @@
+#[allow(unused_field)]
+module dwallet_system::sui_state_proof {
+
+    use dwallet::object::{Self, ID, UID};
+    use dwallet_system::dwallet::{Self, DWalletCap, MessageApproval};
+    use dwallet::tx_context::{TxContext};
+    use dwallet::transfer;
+    use dwallet::bcs;
+    use dwallet::event;
+
+
+
+    struct StateProofRegistry has key, store {
+            id: UID,
+            highest_epoch: u64,
+    }
+
+    struct StateProofConfig has key, store{
+        id: UID,
+        registry_id: ID,
+        package_id: vector<u8>,
+        event_type_layout: vector<u8>,
+        message_field_name: vector<u8>,
+    }
+
+    struct EpochCommittee has key, store {
+        id: UID,
+        committee: vector<u8>,
+    }
+
+
+    struct EpochCommitteeSubmitted has copy, drop {
+        epoch: u64,
+        epoch_committee_id: ID,
+    }
+
+    struct CapWrapper has key, store {
+        id: UID,
+        cap_id_sui: ID,
+        cap: DWalletCap,
+    }
+
+
+    // TODO add wittness here so it can be initialized only once??
+    public fun init_module(init_committee: vector<u8>, package_id: vector<u8>, event_type_layout: vector<u8>, message_field_name: vector<u8>, ctx: &mut TxContext) {
+        let registry = StateProofRegistry {
+            id: object::new(ctx),
+            highest_epoch: 0,
+        };
+
+        let config = StateProofConfig {
+            id: object::new(ctx),
+            registry_id: object::id(&registry),
+            package_id: package_id,
+            event_type_layout: event_type_layout,
+            message_field_name: message_field_name,
+        };
+
+        let first_committee = EpochCommittee {
+            id: object::new(ctx),
+            committee: init_committee,
+        };
+
+        event::emit(EpochCommitteeSubmitted {
+            epoch: 0,
+            epoch_committee_id: object::id(&first_committee),
+        });
+        
+        transfer::share_object(registry);
+        transfer::freeze_object(config);
+        transfer::freeze_object(first_committee);
+    }
+
+    
+
+    native fun sui_state_proof_verify_committee(prev_committee: vector<u8>, checkpoint_summary: vector<u8>): vector<u8>;
+
+
+    public fun submit_new_state_committee(
+        registry: &mut StateProofRegistry,
+        prev_committee: &EpochCommittee,
+        new_checkpoint_summary: vector<u8>,
+        ctx: &mut TxContext,
+    ) {
+        let committee_verified_bytes = sui_state_proof_verify_committee(prev_committee.committee, new_checkpoint_summary);
+
+        let committee_new = EpochCommittee {
+                                    id: object::new(ctx),
+                                    committee: committee_verified_bytes,
+                                    };
+
+        registry.highest_epoch = registry.highest_epoch + 1;
```

Comment:
why is this true?
this proof only passes for sequtnial changes? there's no way to prove an epoch that's further in time than the immediately next epoch?

Reply:
is there no way to get the epoch from the proof anyways? seems like a best practice. Perhaps after that you could assert it is indeed `prev_epoch + 1`

Reply:
In any case this needs proper documentation

Reply:
Here's an attack for you. 
I start in epoch `x`, proving it correctly yields an `EpochCommittee`. 
Now I prove epoch `x + 1` using this `EpochCommittee` object. This increases the registry's highest epoch to `x+1`. It creates an `EpochCommittee` for epoch `x+1`.
Now I submit the same proof again against the old `EpochCommittee` for epoch `x`. This proof verifies, and a new `EpochCommittee` object with the same value is created, again for epoch `x+1`. But now, the registry's highest epoch is set to `x+2`, wheras there's not a single `EpochCommittee` object for that epoch. I can do this indefinetely. Also, I can overflow the u64 potentially (perhaps the fees won't make it possible tho) and even in the same transaction.

Reply:
Notice that in this attack you are emitting events that has a wrong `EpochCommittee` to `epoch` relation

Reply:
This fixes it